<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <script>
      class HttpError extends Error {
        constructor(response) {
          super(`${response.status} for ${response.url}`);
          this.name = 'HttpError';
          this.response = response;
        }
      }

      async function loadJson(url) {
        const response = await fetch(url);

        if (response.status === 200) {
          return response.json();
        }

        throw new HttpError(response);
      }

      // 询问用户名，直到 github 返回一个合法的用户
      async function demoGithubUser() {
        let user;

        while (true) {
          const name = prompt('Enter a name?', 'iliakan');

          try {
            user = await loadJson(`https://api.github.com/users/${name}`);
            break;
          } catch (err) {
            if (err instanceof HttpError && err.response.status === 404) {
              alert('No such user, please reenter.');
            } else {
              throw err;
            }
          }
        }

        alert(`Full name: ${user.name}.`);
        return user;
      }

      demoGithubUser();
    </script>
  </body>
</html>
